libs/Microsoft.MixedReality.WebRTC/PeerConnection.cs

```diff
@@ -1470,10 +1470,6 @@ public void RemoveDataChannel(DataChannel dataChannel)
                 Utils.ThrowOnErrorCode(res);
 
                 DataChannelRemoved?.Invoke(dataChannel);
```

Comment:
If we're relying on `PeerConnection_RemoveDataChannel` to always call internally the interop callback `PeerConnectionInterop.DataChannelRemovedCallback`, we're assuming `OnDataChannelRemoved()` is called which will raise the `DataChannelRemoved` event. So isn't raising the event here too a duplicate?

Reply:
Yes I agree, I wonder why it is done in this manner though. Could it be that the event should always fire independent if it is the peer or the remote that removes it, but that should already be hooked up from the interop callback PerConnectionInterop.DataChannelRemovedCallback?

I've tested some more with the TestNetCoreConsole, starting two instances and trying to close one connection. I added a handler for RemovedDataChannel which just logs to console. Seems like the remote does not notice that the data channels are removed. Seems like there are more problems here!

Reply:
An update, my setup seems to not be working properly, the data channel never goes to Connected. Probably why the client side never knows about it... I will continue investigating tomorrow. It might be a better fix to remove the invokation of the event. That seems to be how the AddDataChannel method is implemented, it does not fire the add event right?

Reply:
Ok, I thought that the DataChannelRemoved event was going to fire when the data channel state changed to closed, but of course that is not the case, that would be up to the client to decide how to handle the data channels.

I'll continue tomorrow!

Reply:
Okay I need help. I'm unsure what this comment means, is it that the removed event is fired only on the local instance of the peer connection, or that it is fired also when the data channel is removed on the remote?
```csharp
        /// <summary>
        /// Event fired when a data channel is removed from the peer connection.
        /// This event is always fired, whatever its creation method (negotiated or not)
        /// and original creator (local or remote peer).
        /// </summary>
        public event DataChannelRemovedDelegate DataChannelRemoved;
```
There might be some hooks missing for this?

Reply:
> I've tried moving the data channel open method and it seems that it works just when I have it after initialize, but not afterwards.

This is a known Google limitation (or maybe even from the WebRTC standard) that data channels need SRTP handshake during initial connection to secure them, and if you don't have any data channel before that initializing then the SRTP handshake is not performed. Later when you add a data channel, it can't be initialized because SRTP is not ready. So you always have to create at least 1 data channel _before_ you initialize, if you ever want to use _any_ data channel during your peer connection lifetime.

> Okay I need help. I'm unsure what this comment means

This is a comment on the C# code so only documents the intended behavior of the C# API itself, unrelated to how the internal underlying Google implementation works. Unfortunately that C++ low-level implementation (`libwebrtc`) is difficult to use and inconsistent at times. I cannot remember on top of my head the details, but I know some events are not fired consistently, so we had to manually craft some in MixedReality-WebRTC to make them more consistent to the end users.

The comment says that we aim at a behavior where the `DataChannelRemoved` event is raised when any data channel is removed, whether it was removed manually by the local peer connection (by an API call), or removed remotely by the other remote peer connection (via a message transmitted over the network). It also says that this behavior is to be consistent between data channels negotiated or not, that is either of:
- the data channels whose ID is pre-agreed by both peers and not transmitted over the network (they both create a data channel with the same ID, that should be available) -- those are out-of-band data channels because the ID is never exchanged, it's pre-agreed upon (_e.g._ "for my app, data channel ID 24 is the text message"). Note that here both peers need to create via API call the data channel for it to be useable.
- the data channels created by one peer without an explicit ID, and the implementation picked an ID and transmitted it over the network to the other peer that then reactively created a data channel with the same ID -- those are in-band data channels because the ID is sent in-band through the peer connection. Note that here only 1 peer creates the data channel via API call; the other peer will auto-create it internally after negotiation, and raise a `DataChannelAdded` event to the app.

All of that however is not very important, the point here is that `libwebrtc` only raise add/remove events in _some_ cases (in general when not from a user-initiated API call), but we aim at having a more consistent API in C# where events are always raised in all cases, so we manually craft the missing ones.

In short, this comment doesn't give you any info on what the internal implementation does, nor on what/where events should be raised from the C# API.

Reply:
Thanks for through explanation!

I dug a bit and found the peerconnectioninterface.h, as well as the datachannelinterface.h, as well as the documentation for webrtclib here https://w3c.github.io/webrtc-pc/#interface-definition.

It looks like there is no implementation for DataChannelRemoved which is called when a remote removes the data channel. Only the state of the data channel will be changed. So what I thought was a bug was just not supported, and doesn't need to be since you can check the data channel state instead.

This said. I've updated the code to take this into consideration. I've changed so we don't call the event like you commented, since it is called by the c wrapper. I've also removed the DataChannels.Remove(dataChannel) call in OnDataChannelRemoved, since that is already done in the remove method.

I will give this a try tomorrow!

Reply:
Ok, I've gone through this once again, when a PeerConnection is closed the data channels are actually removed, and the event is fired, so we need to have the DataChannels.Remove(dataChannel) call in the event handler. It will just return false if the data channel is already removed so no worries.

Can you have another look @djee-ms?

Reply:
I won't have time today sorry, and I'll be away next week unfortunately. Maybe if @fibann has some spare time, otherwise this will have to wait for my return.

Reply:
No worries, I can continue testing and working with packages built from this, so no stress!

Reply:
@djee-ms, I've used this for a while now, seems to be working fine! Hopefully you have some time to look at this! Always feels better to build a nugetpackage from main than a branch:)

Reply:
Ack, and many thanks for your change. I'll try to find some time this week to have a look again at this PR. Sorry for the inconvenience, we're light on time allocated to this project.